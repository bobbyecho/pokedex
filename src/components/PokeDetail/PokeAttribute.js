import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  textValue: {
    fontWeight: 'bold',
    fontSize: 15,
  },
});
export default function DetailAttrPoke({ attribute, value, color }) {
  return (
    <View>
      <Text style={{ color }}>{attribute}</Text>
      <Text style={styles.textValue}>{value}</Text>
    </View>
  );
}
