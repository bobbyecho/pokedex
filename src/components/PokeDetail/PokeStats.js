import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import metrics from '@themes/metrics';
import { capitalize } from '@helpers/transforms';

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: metrics.distance.xl,
  },
  pokeLineContainer: {
    paddingVertical: metrics.distance.m,
  },
  pokeLine: {
    height: 7,
    marginTop: 4,
  },
  lineTitle: {
    fontWeight: '700',
    fontSize: 13,
  },
});

function Line({ name, value, color }) {
  console.log(name, value);
  return (
    <View style={styles.pokeLineContainer}>
      <Text style={styles.lineTitle}>{capitalize(name)}</Text>
      <View
        style={[
          styles.pokeLine,
          {
            width: (metrics.screen.fullWidth * value) / 265,
            backgroundColor: color,
          },
        ]}
      />
    </View>
  );
}

export default function ({ stats, color }) {
  return (
    <View style={styles.container}>
      {stats.map((v) => {
        const {
          stat: { name },
          base_stat: baseStat,
        } = v;

        return <Line key={name} name={name} value={baseStat} color={color} />;
      })}
    </View>
  );
}
