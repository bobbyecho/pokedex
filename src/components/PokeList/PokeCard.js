import React from 'react';
import { useNavigation } from '@react-navigation/native';
import { View, Image, StyleSheet } from 'react-native';
import { Caption, Text, TouchableRipple } from 'react-native-paper';
import { pokeImgUrl, capitalize } from '@helpers/transforms';
import metrics from '@themes/metrics';
import colors from '@themes/colors';
import PokeElementalIcon from '@components/PokeList/PokeElementalIcon';

const styles = StyleSheet.create({
  pokeCard: {
    alignItems: 'center',
    flexDirection: 'row',
    margin: metrics.distance.xs,
    backgroundColor: colors.white,
    borderRadius: metrics.radius.commonRadius,
  },
  pokeImage: {
    flex: 1,
    width: '100%',
    height: 100,
    resizeMode: 'contain',
  },
  imageContainer: {
    padding: metrics.distance.s,
    width: '40%',
    height: '100%',
  },
  pokeCardContent: {
    flex: 1,
    alignContent: 'center',
    paddingLeft: metrics.distance.xl,
    width: '60%',
  },
  pokeName: {
    fontWeight: 'bold',
    fontSize: 32,
  },
  pokeType: {
    fontSize: 15,
    marginLeft: metrics.distance.s,
  },
  pokeElement: {
    flexDirection: 'row',
    paddingVertical: metrics.distance.s,
  },
});

export default ({ data }) => {
  const navigation = useNavigation();
  const { name, types } = data;
  const pokeType = types[0].type.name;

  function goToDetail() {
    navigation.navigate('PokeDetail', { pokeData: data });
  }

  return (
    <TouchableRipple onPress={goToDetail}>
      <View style={styles.pokeCard}>
        <View style={styles.pokeCardContent}>
          <Text style={[styles.pokeName, { color: colors[pokeType] }]}>
            {capitalize(name)}
          </Text>
          <View style={styles.pokeElement}>
            <PokeElementalIcon name={pokeType} />
            <Caption style={styles.pokeType}>{pokeType}</Caption>
          </View>
        </View>
        <View style={styles.imageContainer}>
          <Image
            source={{ uri: pokeImgUrl(name) }}
            style={styles.pokeImage}
            onL
          />
        </View>
      </View>
    </TouchableRipple>
  );
};
