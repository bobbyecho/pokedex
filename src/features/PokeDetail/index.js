import React from 'react';
import { StyleSheet, View } from 'react-native';
import Container from '@components/Container';
import metrics from '@themes/metrics';
import colors from '@themes/colors';
import PokeCard from '@components/PokeList/PokeCard';
import PokeAttribute from '@components/PokeDetail/PokeAttribute';
import PokeStats from '@components/PokeDetail/PokeStats';
import { capitalize } from '@helpers/transforms';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.white,
    height: metrics.screen.fullHeight,
    width: metrics.screen.fullWidth,
    justifyContent: 'center',
  },
  pokeDetailContainer: {
    borderRadius: metrics.radius.commonRadius,
    backgroundColor: colors.white,
    margin: metrics.distance.xxl,
    paddingVertical: metrics.distance.m,
  },
  pokeAttributes: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    paddingVertical: metrics.distance.m,
  },
});

function PokeDetail(props) {
  const {
    route: {
      params: { pokeData },
    },
  } = props;

  const pokeType = pokeData.types[0].type.name;
  const pokeHeight = pokeData.height;
  const pokeWeight = pokeData.weight;
  const pokeAbility = pokeData.abilities[0].ability.name;
  const themeColor = colors[pokeType];
  const pokeStats = pokeData.stats;

  return (
    <Container style={[styles.container, { backgroundColor: themeColor }]}>
      <View style={styles.pokeDetailContainer}>
        <PokeCard data={pokeData} />
        <View style={styles.pokeAttributes}>
          <PokeAttribute
            attribute="Height"
            value={`${pokeHeight}"`}
            color={themeColor}
          />
          <PokeAttribute
            attribute="Weight"
            value={`${pokeWeight}lbs`}
            color={themeColor}
          />
          <PokeAttribute
            attribute="Ability"
            value={capitalize(pokeAbility)}
            color={themeColor}
          />
        </View>
        <PokeStats stats={pokeStats} color={themeColor} />
      </View>
    </Container>
  );
}

export default PokeDetail;
