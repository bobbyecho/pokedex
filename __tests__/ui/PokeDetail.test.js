import React from 'react';
import { Text } from 'react-native';
import PokeDetail from '@features/PokeDetail';
import mockListPokemon from '../../__mocks__/payload/listPokemon';
import renderer from 'react-test-renderer';
import PokeCard from '@components/PokeList/PokeCard';

describe('Pokemon Detail', () => {
  let container;
  let navParam = {
    route: {
      params: {
        pokeData: mockListPokemon[0],
      },
    },
  };

  container = renderer.create(<PokeDetail {...navParam} />);
  const instance = container.root;

  it('should render successfully', () => {
    expect(container.toJSON()).toMatchSnapshot();
  });

  it('should have pokemon name "ditto"', () => {
    const PokeCardInstance = instance.findByType(PokeCard);
    const TextInstance = PokeCardInstance.findAllByType(Text);
    expect(TextInstance[0].props.children).toBe('Ditto');
  });
});
