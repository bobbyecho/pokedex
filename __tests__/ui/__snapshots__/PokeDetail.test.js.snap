// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Detail should render successfully 1`] = `
<View
  showsVerticalScrollIndicator={null}
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "height": 1334,
        "justifyContent": "center",
        "width": 750,
      },
      Object {
        "backgroundColor": "#A8A878",
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 4,
        "margin": 24,
        "paddingVertical": 12,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          false,
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 4,
            "flexDirection": "row",
            "margin": 4,
          }
        }
      >
        <View
          style={
            Object {
              "alignContent": "center",
              "flex": 1,
              "paddingLeft": 20,
              "width": "60%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 32,
                    "fontWeight": "bold",
                  },
                  Object {
                    "color": "#A8A878",
                  },
                ],
              ]
            }
          >
            Ditto
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
                "paddingVertical": 8,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/images/poke-elements/normal.png",
                }
              }
              style={
                Object {
                  "height": 24,
                  "width": 24,
                }
              }
            />
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "color": "rgba(0, 0, 0, 0.54)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 12,
                        "letterSpacing": 0.4,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      Object {
                        "fontSize": 15,
                        "marginLeft": 8,
                      },
                    ],
                    Array [
                      Object {
                        "fontSize": 12,
                        "letterSpacing": 0.4,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      Object {
                        "fontSize": 15,
                        "marginLeft": 8,
                      },
                    ],
                  ],
                ]
              }
            >
              normal
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "height": "100%",
              "padding": 8,
              "width": "40%",
            }
          }
        >
          <Image
            onL={true}
            source={
              Object {
                "uri": "https://img.pokemondb.net/artwork/ditto.jpg",
              }
            }
            style={
              Object {
                "flex": 1,
                "height": 100,
                "resizeMode": "contain",
                "width": "100%",
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-around",
          "paddingVertical": 12,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#A8A878",
            }
          }
        >
          Height
        </Text>
        <Text
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
            }
          }
        >
          3"
        </Text>
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#A8A878",
            }
          }
        >
          Weight
        </Text>
        <Text
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
            }
          }
        >
          40lbs
        </Text>
      </View>
      <View>
        <Text
          style={
            Object {
              "color": "#A8A878",
            }
          }
        >
          Ability
        </Text>
        <Text
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
            }
          }
        >
          Imposter
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "paddingHorizontal": 20,
        }
      }
    >
      <View
        style={
          Object {
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 13,
              "fontWeight": "700",
            }
          }
        >
          Speed
        </Text>
        <View
          style={
            Array [
              Object {
                "height": 7,
                "marginTop": 4,
              },
              Object {
                "backgroundColor": "#A8A878",
                "width": 135.8490566037736,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 13,
              "fontWeight": "700",
            }
          }
        >
          Special-defense
        </Text>
        <View
          style={
            Array [
              Object {
                "height": 7,
                "marginTop": 4,
              },
              Object {
                "backgroundColor": "#A8A878",
                "width": 135.8490566037736,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 13,
              "fontWeight": "700",
            }
          }
        >
          Special-attack
        </Text>
        <View
          style={
            Array [
              Object {
                "height": 7,
                "marginTop": 4,
              },
              Object {
                "backgroundColor": "#A8A878",
                "width": 135.8490566037736,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 13,
              "fontWeight": "700",
            }
          }
        >
          Defense
        </Text>
        <View
          style={
            Array [
              Object {
                "height": 7,
                "marginTop": 4,
              },
              Object {
                "backgroundColor": "#A8A878",
                "width": 135.8490566037736,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 13,
              "fontWeight": "700",
            }
          }
        >
          Attack
        </Text>
        <View
          style={
            Array [
              Object {
                "height": 7,
                "marginTop": 4,
              },
              Object {
                "backgroundColor": "#A8A878",
                "width": 135.8490566037736,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 13,
              "fontWeight": "700",
            }
          }
        >
          Hp
        </Text>
        <View
          style={
            Array [
              Object {
                "height": 7,
                "marginTop": 4,
              },
              Object {
                "backgroundColor": "#A8A878",
                "width": 135.8490566037736,
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
